### ISSUE TITLE
Module:write prevents a File:writeObject hook from propagating

### ISSUE THREAD

**Issue by genixpro** [torch/nn#847]
I'm using the Torch.file class and trying to write out an object, like so:

```
    -- This function is used to convert all the tensors written out 
    -- into regular CPU tensors. Rather then double tensors
    local convertCPUTensor = function(object)
        if torch.typename(object) == 'torch.ClTensor' then
            return object:double()
        end
        return object
    end

    -- Assemble the model data into one object
    local modelData = {
        modelOptions = self.modelOptions,
        neuralNetwork = self.neuralNetwork
    }

    local file = torch.DiskFile(outputFilename, 'w')
    file:writeObject(modelData, '', convertCPUTensor)
    file:close()
```

I'm using the 'hook' parameter that writeObject accepts in order to make sure ALL gpu tensors get converted into regular CPU tensors. Before you ask, I have used self.neuralNetwork:double(), module:double(), and similar functions. These functions are SUPPOSED to work, but due to some bug one of the torch libraries that I was unable to track down, every time I wrote my models out to the disk, there was always four remaining GPU tensors, even after applying node:double() on every single node I could get my hands on. This made it impossible to open up and use these models on machines which didn't have OpenCL installed. Thus, I had to implement this hack in order to ensure that every single GPU tensor gets converted into a regular CPU tensor.
 
HOWEVER! The hack gets worse. Because of the Module:write function, my hook doesn't propagate all the way down, instead it gets forgotten when File.writeObject calls the Module:write metamethod at https://github.com/torch/torch7/blob/master/File.lua#L210. 

Strangely, I found that Module:write and Module:read are completely superfluous. If I just delete those functions entirely using the following hack:

```
    -- This is a bit of a hack. If we don't do this, then our convertCPUTensor hook
    -- doesn't get recursed all the way down in the save function, and there will
    -- be a few left over tensors that never get converted. Surprisingly this works
    local meta = torch.getmetatable("nn.Module")
    meta.write = nil
    meta.read = nil
```

And everything works great! It seems like these methods are totally superfluous and actually makes using that hook parameter impossible.

There is of course a second bug, why did using node:double() not convert every single GPU tensor to a CPU tensor in the first place (which would have prevented the need for these hacks in the first place) but I don't actually know which library that bug is located in. It could be torch7, nn, nngraph or rnn.

**Comment by fmassa** [comment]
Thanks for pointing out this issue. Indeed it seems that we can't pass the `hook` function to `nn` modules as you pointed out. Indeed, the `read` and `write` methods of `nn.Module` [don't do much more than simply copying the module](https://github.com/torch/nn/blob/master/Module.lua#L148-L162). But there are custom modules that overwrite those methods, like [`DataParallelTable` in `cunn`](https://github.com/torch/cunn/blob/master/DataParallelTable.lua#L434-L455), so erasing the `read` and `write` fields is not a solution.

We might consider adding the `hook` field to `nn.Module.write` though, [as it was recently introduced to torch](https://github.com/torch/torch7/pull/567), @soumith what do you think?

It seems to me that the main problem here is that there is a bug in the `type` function somewhere. Could you send your model definition with a test case illustrating this conversion issue? Also, does this problem only happen when converting from `cltensor` to `double`, or does it happen when converting from `cuda` to `double` ?

**Comment by genixpro** [comment]
I have not used CUDA with Torch yet, only OpenCL. There would seem to be a bug in the type function, or in how it interacts with the Torch rnn library https://github.com/Element-Research/rnn

**Comment by nicholas-leonard** [comment]
@genixpro Can you copy-paste your model? Maybe that will give us a clue to where the conversion isn't happening.

**Comment by genixpro** [comment]
I will as best as I can. The code is a little convoluted, so I'll try to cut down the code to only the relevant pieces:
`
function ReceiptModelTrainer:createModel()
        self.neuralNetwork = {}
        self.lstmLayers = {}
        self.lastLayerOutputSize = self.inputSize

        self.neuralNetwork.sequential_rnn = nn.Identity()
        for layer=1,self.modelOptions.num_layers do
            self:addLSTMLayer()
        end

        self.neuralNetwork.softmax = nn.Sequential()
        self.neuralNetwork.softmax:add(nn.Linear(self.lastLayerOutputSize, self.outputSize)):add(nn.LogSoftMax())
        self.neuralNetwork.sequential_softmax = nn.Sequencer(self.neuralNetwork.softmax)

        self.neuralNetwork.criterion = nn.ClassNLLCriterion()
        self.neuralNetwork.sequential_criterion = nn.SequencerCriterion(self.neuralNetwork.criterion)
end
`

`
function ReceiptModelTrainer:addLSTMLayer()
    -- This represents the input to the LSTM as a whole
    local lstmInput = nn.Identity()()

    local layerNumber = #self.lstmLayers + 1

    -- Holds the LSTM and an optional dropout in sequence
    local layerSequential = nn.Sequential()

    -- Create the actual LSTM node itself, taken from the RNN library
    local newLSTMLayer = nn.SeqLSTM(self.lastLayerOutputSize, self.modelOptions.rnn_size)

    local newLSTMParams, newLSTMGrads = newLSTMLayer:parameters()
    -- Reset the parameters, including setting the forget-gate biases if thats desired
    newLSTMParams[1]:uniform(-0.08, 0.08)
    newLSTMParams[2]:uniform(-0.08, 0.08)
    self:resetLSTMLayerForgetBias(newLSTMLayer)

    -- Gradients for the new layer must be zero, since its just starting
    newLSTMGrads[1]:zero()
    newLSTMGrads[2]:zero()

    -- Add the new lstm layer to the sequential node for the layer
    table.insert(self.lstmLayers, newLSTMLayer)
    layerSequential:add(newLSTMLayer)

    -- Add in a dropout
    if self.modelOptions.dropout > 0 and (self.modelOptions.enable_internal_dropout == 1 or layerNumber == self.modelOptions.num_layers) then
        layerSequential:add(nn.Sequencer(nn.Dropout(self.modelOptions.dropout)))
    end

    local existingNetwork = self.neuralNetwork.sequential_rnn(lstmInput)
    self.neuralNetwork.sequential_rnn = nn.gModule({lstmInput}, {layerSequential(existingNetwork)})

    self.lastLayerOutputSize = self.modelOptions.rnn_size
end
`

Notice that the addLSTMLayer function is essentially wrapping whatever the current network is with a new network which takes the existing output and sends it through another LSTM layer. This is done because it allowed us to easily add in another layer during training. 

I should note that I've seen this issue with other formulations of the model as well, including when we did our entire own LSTM model which was modelled after char-rnn at https://github.com/karpathy/char-rnn/blob/master/model/LSTM.lua

**Comment by soumith** [comment]
@genixpro i've fixed the bug in the nngraph package that affects typecasting all tensors in the graph. https://github.com/torch/nngraph/pull/126

Please reinstall nngraph (luarocks install nngraph) and give it a shot.
Thanks for reporting the bug and taking the time to give a test case.
